### PR TITLE
image encoding as a standalone function

### PR DESCRIPTION
**Pull request by bbferka**
Small improvement to your PR, making the compression of images into a stand alone function as well. As a reference ``compressed_depth_image_transport`` also has it similar to this.  I tested this using the usb_cam package.

### CODE REVIEW

File: compressed_image_transport/src/compression_common.cpp [hunk 1]
```diff
@@ -37,6 +37,7 @@
 #include <cv_bridge/cv_bridge.h>
 #include <opencv2/imgproc/imgproc.hpp>
 #include <opencv2/highgui/highgui.hpp>
+#include <ros/ros.h>
```
Comment (airballking) [inline compressed_image_transport/src/compression_common.cpp:40]: Bigger change request coming in. Sorry! ;) 

I removed all dependencies on ROS logging such that decoding can happen without a rosnode. I think that is an important feature, and the same applies to the encoding. Could you please replace all logging calls either with exceptions, printouts, or comments, where appropriate and necessary?

File: compressed_image_transport/src/compressed_publisher.cpp [hunk 2]
```diff
@@ -112,122 +103,26 @@ void CompressedPublisher::publish(const sensor_msgs::Image& message, const Publi
...
+      compressed = encodeImage(message, encodingFormat, params);
       break;
     }
-      // PNG Compression
+
+    // PNG Compression
     case PNG:
     {
-      params.resize(3, 0);
+      params.resize(2, 0);
```
Comment (airballking) [inline compressed_image_transport/src/compressed_publisher.cpp:125]: Why did you change that magic number?

File: compressed_image_transport/src/compressed_publisher.cpp [hunk 3]
```diff
@@ -78,31 +78,22 @@ void CompressedPublisher::configCb(Config& config, uint32_t level)
...
+  // Compression settings
+  std::vector<int> params;
+  sensor_msgs::CompressedImagePtr compressed;
   switch (encodingFormat)
   {
     // JPEG Compression
     case JPEG:
     {
-      params.resize(9, 0);
+      params.resize(8, 0);
```
Comment (airballking) [inline compressed_image_transport/src/compressed_publisher.cpp:96]: Also here, why did that number change?

**Review by airballking — CHANGES_REQUESTED** [review]
Thank you for your contribution! I agree that it is a good idea to add this to the pull request upstream. I requested some changes, could you please look into them?

File: compressed_image_transport/src/compressed_publisher.cpp (hunk 3 above)
Comment (bbferka) [reply compressed_image_transport/src/compressed_publisher.cpp:96]: It is the size of the local vector. Which is eight and not nine if you look at the lines below where the vector is filled up; Furthermore it is just plain wrong to have an odd number as the size of this container since it needs to have pairs of OpenCV flags and values. As stated by the documentation the params vector contains _"Format-specific parameters encoded as pairs"_; (see docu [here](https://docs.opencv.org/3.4/d4/da8/group__imgcodecs.html#gabbc7ef1aa2edfaa87772f1202d67e0ce) )

File: compressed_image_transport/src/compressed_publisher.cpp (hunk 2 above)
Comment (bbferka) [reply compressed_image_transport/src/compressed_publisher.cpp:125]: see below;

File: compressed_image_transport/src/compression_common.cpp (hunk 1 above)
Comment (bbferka) [reply compressed_image_transport/src/compression_common.cpp:40]: Makes sense. I'll change this and commit soon; But, ROS_INFO or ROS_DEBUG do not need need a node to work; I never though about this actually so quickly had to test it. A simple hello world with only a ROS_INFO will confirm what I just wrote. The ROS_ERROR messages I can easily replace with throwing exceptions and catching them in the publisher but the ROS_DEBUG is trickier; replacing it with std::cout or std::cerr will not have the same behaviour as was intended in the original code that I'm refactoring here.

**Comment by bbferka** [comment]
@airballking ping!

**Comment by bbferka** [comment]
@airballking pong!

File: compressed_image_transport/src/compression_common.cpp (hunk 1 above)
Comment (airballking) [reply compressed_image_transport/src/compression_common.cpp:40]: Oh, interesting! I did not know that ROS_INFO and ROS_DEBUG work without roscore.

Replacing ROS_DEBUG is indeed an issue. If it works without roscore, I'd be happy with just leaving it in. That would mean that only the ROS_ERROR calls need to be replaced by throwing exceptions. Is that already in?

File: compressed_image_transport/src/compressed_publisher.cpp (hunk 3 above)
Comment (airballking) [reply compressed_image_transport/src/compressed_publisher.cpp:96]: OK. So, also a bug-fix. Good.

File: compressed_image_transport/src/compressed_publisher.cpp (hunk 2 above)
Comment (airballking) [reply compressed_image_transport/src/compressed_publisher.cpp:125]: Also good.

File: compressed_image_transport/src/compression_common.cpp (hunk 1 above)
Comment (airballking) [reply compressed_image_transport/src/compression_common.cpp:40]: Just saw that the newest commit replaces the ROS_ERROR calls with throwing exceptions. Solved, then!

**Comment by airballking** [comment]
@bbferka Thanks for the improvements. Will merge now. Let's see what it does to the upstream PR.
